src/wallet-redux.js

```diff
@@ -22,137 +24,246 @@ export const WalletType = {
+                    wallets.set(
+                        wallet.accountId,
+                        Object.assign({}, wallets.get(wallet.accountId), entry)
+                    );
+                }
+
+                wallet.addresses.forEach(address => {
+                    const entry = {
+                        address: address.address,
+                        label: address.label,
+                        type: wallet.type === WalletType.LEDGER ? AccountType.LEDGER : AccountType.KEYGUARD_HIGH,
+                        isLegacy: wallet.type === WalletType.LEGACY,
+                        walletId: wallet.accountId,
+                    };
+
+                    // merge with previous information
+                    accounts.set(
+                        address.address,
+                        Object.assign({}, accounts.get(address.address), entry)
+                    );
+                });
+
+                wallet.contracts.forEach(contract => {
+                    const entry = Object.assign({}, contract, {
+                        type: contract.type === 1 /* Nimiq.Account.Type.VESTING */ ? AccountType.VESTING : AccountType.HTLC,
+                        stepAmount: contract.stepAmount / 1e5,
+                        totalAmount: contract.totalAmount / 1e5,
+                        label: contract.label,
+                        owner: contract.owner, // vesting only
+                        sender: contract.sender, // HTLC only
+                        recipient: contract.recipient, // HTLC only
+                        isLegacy: wallet.type === WalletType.LEGACY,
+                        walletId: wallet.accountId,
+                    });
+
+                    // merge with previous information
+                    accounts.set(
+                        contract.address,
+                        Object.assign({}, accounts.get(contract.address), entry)
+                    );
+                });
             });
 
-            const newState = Object.assign({}, state, {
+            return updateState({
                 hasContent: true,
-                // converts array to map with id as key
-                entries: new Map(newEntries)
+                wallets,
+                accounts,
             });
+        }
 
-            const activeWalletIsEmptyLegacy = state.activeWalletId === LEGACY && !action.hasLegacyAccounts;
-            const activeWalletWasDeleted = state.activeWalletId !== LEGACY && !newState.entries.get(state.activeWalletId);
+        case TypeKeys.REMOVE_ACCOUNT: {
+            const accounts = new Map(state.accounts);
+            accounts.delete(action.address);
 
-            if (activeWalletIsEmptyLegacy || activeWalletWasDeleted) {
-                newState.activeWalletId = action.walletWithMostAccounts.id;
-            }
+            const accountArray = [...accounts.values()];
 
-            return newState;
+            const affectedVestingContracts = accountArray.filter(account =>
```

Comment:
Instead of doing the checks for the contracts again, couldn't you just use the `addressesToKeep` to delete all accounts & contracts which are not in that list. Then you would only have to do the filtering once.